### ISSUE TITLE
Initialize has not been called on this module yet

### ISSUE THREAD

**Issue by haldiggs** [Microsoft/ApplicationInsights-dotnet-server#381]
recently we started seeing this exception being raised when an application is first being loaded after an appPool cycle. Has something changed with the telemetryClient init?

    public void OnError(HttpContext context)
    {
            if (this.telemetryClient == null)
            {
                throw new InvalidOperationException("Initialize has not been called on this module yet.");
            }
shouldn't this thing get initialized somewhere rather than just throwing an exception? It doesn't seem like I have anyway to deal with this except to tap the site a 2nd time and tell my users to deal with it.

**Comment by SergeyKanzhelev** [comment]
My guess is that this may happen when the initialization of any other module has failed before the module initialization. Please check the telemetry - look for the traces with the "AI:" prefix indicating configuration error.

We should definitely fix the logic to not throw in this case. It is a clear violation of SDK error handling policy

**Comment by lmolkova** [comment]
@SergeyKanzhelev it affects also feature I'm working on.
What is the most appropriate thing to do? Write error to event log or just silently swallow it?

**Comment by SergeyKanzhelev** [comment]
@lmolkova are you talking about `PreRequestHandler` callback? I'd say to trace `Error` at first so it will be visible for customer by default and then `Verbose` without crashing the request.

**Comment by haldiggs** [comment]
but the bigger issue is that this is throwing errors into the event log and there is nothing I can do to recover from it. Why the object has not been created? Its not as if there is anything I can change/call from within my application is there?

This is just odd because its being thrown extremely often and causing my IT department headaches. They see this as an IIS crash issue. Not really a good thing.

**Comment by SergeyKanzhelev** [comment]
@haldiggs I understand that the issue is critical and we will try to at least remove throwing logic as soon as we can. 

As a workaround - have you ensured that the `ApplicationInsights.config` is correct and there are no errors reported by Application Insights SDK? You can also try to run this code manually in Global.asax:

``` csharp
foreach (var module in Microsoft.ApplicationInsights.Extensibility.Implementation.TelemetryModules.Instance.Modules)
{
  module.Initializer(TelemetryConfiguration.Active);
}
```

I'm sorry for the headaches this issue causing.

**Comment by haldiggs** [comment]
thanks for the work around

**Comment by SergeyKanzhelev** [comment]
@haldiggs, can you please confirm that the work around help. We will definitely fix it in upcoming versions

**Comment by haldiggs** [comment]
I wish I could right now. The issue cropped up in a production environment whose primary usage is serving WCF/WAS. ApplicationInsights was updated during the last deployment. 

I am stuck with quarterly deployments in that environment except in emergencies. I ended up having to shut off ApplicationInsights for now. I plan on testing the work around in the development environment but I have to get it into the pipeline. 

For now just knowing there is a possible solution helps. I will report back if testing is fruitful but I'd rather just see it not throw the error and create what it needs as it used to.

Thanks very much for the attention though.

**Comment by haldiggs** [comment]
ok I finally got around to trying a complete removal and replacement of ApplicationInsights using 2.3.0 
![image](https://cloud.githubusercontent.com/assets/1693010/25534668/1f8e0160-2bfa-11e7-8548-a7a2acf38dde.png) and right out of the box I got the same error. I see some work was done based on thsi issue, would that fix be in the pre-release?

**Comment by haldiggs** [comment]
@SergeyKanzhelev regarding the work around above I had to rename "Initializer" to "Initializer" and I added it to Application_Start but it never found any modules to initialize. I have verified that data is being sent. Guessing it might be the original InstrumentationKey so next week I'll create all new and see how that goes.

`		foreach (var module in Microsoft.ApplicationInsights.Extensibility.Implementation.TelemetryModules.Instance.Modules)
			{
    			module.Initialize(Microsoft.ApplicationInsights.Extensibility.TelemetryConfiguration.Active);
			}
`

**Comment by Dmitry-Matveev** [comment]
@haldiggs , did you have a chance to continue on this?

**Comment by haldiggs** [comment]
@Dmitry-Matveev Thanks for keeping up. I was never able to get the first work-around to function properly. I ended up abandoning ApplicationInsights completely until recently. I tried again w/2.4.0.0, this allowed the applicationinsights.config file to be properly constructed when I re-added the packages back again. There were a number of differences so I cannot tell you which one fixed the issue. We are moving forward w/2.4.0.0 at this time.
